### PR TITLE
Deprecate TimeUnit2

### PR DESCRIPTION
**Pull request by batmat**
`java.util.concurrent.TimeUnit` is preferrable now.

Quoting Stephen:
> Java 5 did not have all the units required. So TU2 was one that had better conversion. Can be deprecated 
once we upgrade to Java 6 ;-)

# Description

* I didn't file a JENKINS issue since this is a pretty internal/dev oriented change. But if this is judged necessary, please tell me so and I'll do it.

Details: see title, very small change. And 0 risk from a runtime perspective. 

### Changelog entries

Proposed changelog entries:

_N/A_ I don't even think this should have an entry in the changelog, should it?

### Submitter checklist

- ~~[ ] JIRA issue is well described~~
- ~~[ ] Link to JIRA ticket in description, if appropriate~~
- [X] Appropriate autotests or explanation to why this change has no tests
- ~~[ ] For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc)~~

### Desired reviewers

@reviewbybees

### CODE REVIEW

**Review by stephenc — APPROVED** [review]

**Review by oleg-nenashev — APPROVED** [review]
:bee:

**Comment by oleg-nenashev** [comment]
Maybe `@Restricted(DoNotUse.class)` will be a good addition

**Comment by batmat** [comment]
@stephenc @oleg-nenashev `@Restricted(DoNotUse.class)` also added as requested/proposed.

**Review by stephenc — APPROVED** [review]

**Review by oleg-nenashev — APPROVED** [review]

**Comment by batmat** [comment]
@reviewbybees done

**Review by jglick — CHANGES_REQUESTED** [review]
There are dozens of usages of `TimeUnit2` in this repository. You should clean them up. (I know they are harmless, but as a matter of policy we should not be deprecating things with a claim that a full replacement is available, without proof that this is actually so in some representative code base.)

**Comment by batmat** [comment]
Agreed @jglick I actually have that change locally, but it fails for now because accmod either is limited or, more probably IIUC, has a bug wrt to enums.

```
[INFO] --- access-modifier-checker:1.11:enforce (default) @ jenkins-core ---
[ERROR] hudson/util/TimeUnit2$1:0 hudson/util/TimeUnit2 must not be used
[ERROR] hudson/util/TimeUnit2$2:0 hudson/util/TimeUnit2 must not be used
[ERROR] hudson/util/TimeUnit2$7:0 hudson/util/TimeUnit2 must not be used
[ERROR] hudson/util/TimeUnit2$4:0 hudson/util/TimeUnit2 must not be used
[ERROR] hudson/util/TimeUnit2$3:0 hudson/util/TimeUnit2 must not be used
[ERROR] hudson/util/TimeUnit2$6:0 hudson/util/TimeUnit2 must not be used
[ERROR] hudson/util/TimeUnit2$5:0 hudson/util/TimeUnit2 must not be used
```

**Comment by jglick** [comment]
Arguably a bug in `access-modifier-checker`. You could fix that and `deploy` a snapshot for use in this PR, or just relax to `NoExternalUse` for now.

**Comment by batmat** [comment]
@jglick fixed, build is passing on my machine. 
It's for now using a SNAPSHOT of https://github.com/kohsuke/access-modifier/issues/6 while it's not merged/released.

File: core/src/main/java/hudson/util/TimeUnit2.java [hunk 1]
```diff
@@ -63,7 +67,12 @@
  *
  * @since 1.5
  * @author Doug Lea
+ * @deprecated use {@link TimeUnit}. (Java 5 did not have all the units required, so TimeUnit2 was introduced because it
+ * had better conversion until Java 6 went out.)
  */
+@Deprecated
+@RestrictedSince("TODO")
+@Restricted(DoNotUse.class)
```
Comment (jglick) [inline core/src/main/java/hudson/util/TimeUnit2.java]: Do we actually need this to be `@Restricted`? Why? `@Deprecated` accomplishes everything we need, I think.

File: pom.xml [hunk 2]
```diff
@@ -95,7 +95,7 @@ THE SOFTWARE.
     <findbugs-maven-plugin.version>3.0.4</findbugs-maven-plugin.version>
     <findbugs.failOnError>true</findbugs.failOnError>
     <test-annotations.version>1.2</test-annotations.version>
-    <access-modifier.version>1.11</access-modifier.version>
+    <access-modifier.version>1.12-PR6-SNAPSHOT</access-modifier.version>
```
Comment (jglick) [inline pom.xml]: You need to change the two properties below instead of this one.

File: core/src/main/java/hudson/util/TimeUnit2.java (hunk 1 above)
Comment (batmat) [inline core/src/main/java/hudson/util/TimeUnit2.java]: It was recommended by @oleg-nenashev and encouraged by @stephenc I guess this avoids we possibly reintroduce some usages as we go @jglick ?

Comment (jglick) [inline core/src/main/java/hudson/util/TimeUnit2.java]: I generally reserve restriction for newly introduced (or modified) APIs, or cases where ongoing use is clearly going to break, so this seems like overkill. I am not objecting, just FYI.

**Comment by batmat** [comment]
[Finally followed the recommendation of @jglick above](https://github.com/jenkinsci/jenkins/pull/2892#issuecomment-302710465):

> [...] just relax to NoExternalUse for now.

In the end, the goal of actually deprecating `TimeUnit2` and that people/code start adapting and replace their usage would be moving forward, which was the gist of the PR...

Sorry for the long-lived PRs :-(

File: core/src/main/java/hudson/util/TimeUnit2.java [hunk 3]
```diff
@@ -63,7 +67,12 @@
  *
  * @since 1.5
  * @author Doug Lea
+ * @deprecated use {@link TimeUnit}. (Java 5 did not have all the units required, so TimeUnit2 was introduced because it
+ * had better conversion until Java 6 went out.)
  */
+@Deprecated
+@RestrictedSince("TODO")
+@Restricted(NoExternalUse.class)
```
Comment (jglick) [inline core/src/main/java/hudson/util/TimeUnit2.java]: A restriction is inappropriate here. `@Deprecated` is appropriate and suffices. The reason for the existence of `@Restricted(NoExternalUse.class)` is to allow access from another package within the same source root, similar to `public` in a nonexported package in Java 9.

File: core/src/main/java/hudson/util/TimeUnit2.java [hunk 4]
```diff
@@ -63,7 +67,12 @@
  *
  * @since 1.5
  * @author Doug Lea
+ * @deprecated use {@link TimeUnit}. (Java 5 did not have all the units required, so TimeUnit2 was introduced because it
```
Comment (jglick) [inline core/src/main/java/hudson/util/TimeUnit2.java]: `@code` or `@link`

**Review by jglick — APPROVED** [review]

**Comment by batmat** [comment]
@reviewbybees done

**Comment by batmat** [comment]
Thanks for the reviews!

**Comment by daniel-beck** [comment]
Since `TimeUnit2` is used in plenty of plugins, and newly restricted core API, a changelog entry is definitely needed.

**Comment by batmat** [comment]
@daniel-beck 

* `hudson.util.TimeUnit2` deprecated, use `java.util.concurrent.TimeUnit` instead. 

?

**Pull request merged**